Shard Split : Add / Update logging

niupre opened the pull request:
DESCRIPTION: Improve logging during shard split and resource cleanup

### DESCRIPTION

This PR makes logging improvements to Shard Split : 

1. Update confusing logging to fix #6312
2. Added new `ereport(LOG` to make debugging easier as part of telemetry review.

### UPDATE
I have removed all connection reuse changes and will make them together as part of updating PR #6314.

Fixes #6312

marcocitus commented on src/backend/distributed/operations/shard_split.c:
```diff
@@ -2238,7 +2291,8 @@ GetNextShardIdForSplitChild()
 
 	shardId = SafeStringToUint64(PQgetvalue(result, 0, 0 /* nodeId column*/));
-	CloseConnection(connection);
+	PQclear(result);
```
now that we no longer close the connection, we need to be careful that we don't cache it for the lifetime of the session.

I suggest introducing a GetLocalConnectionForSubtransactions that opens a connection to localhost with OUTSIDE_TRANSACTION and sessionLifespan set to false.

niupre replied:
Ack. Will do in another PR.

niupre replied:
@marcocitus , I don't see any session flag in 'GetNodeUserDatabaseConnection'. The only thing I could find was 'start_session_level_connection_to_node' matching 'sessionLifespan'. What exactly are you referring to here?

niupre replied:
Discussed offline. This has been renamed to 'forceCloseAtTransactionEnd'.

marcocitus commented on src/backend/distributed/operations/shard_cleaner.c:
```diff
@@ -727,17 +742,22 @@ TryLockRelationAndPlacementCleanup(Oid relationId, LOCKMODE lockmode)
+							 nodePort,
+							 operation)));
+	}
 
```
shouldn't we be fixing this function for connection usage as well?

It may be useful to have a ExecuteCommandListInLocalSubtransaction that does the right thing.

niupre replied:
The changes will be part of other PR as discussed.

niupre replied:
Will do as part of other PR.

marcocitus commented on src/backend/distributed/operations/worker_split_copy_udf.c:
```diff
@@ -113,6 +123,48 @@ worker_split_copy(PG_FUNCTION_ARGS)
+
+	appendStringInfo(splitCopyTrace, "]");
+
+	ereport(LOG, (errmsg("%s", splitCopyTrace->data)));
```
maybe move the above into a separate function that is used in the errmsg part of the ereport. That way, it does not get called if the log level is higher.

niupre replied:
Thanks. Done.

marcocitus reviewed (commented):
On 2 nodes, a simple create_distributed_table_concurrently with 2 nodes opens uses 284 separate connections on main, 252 connections after this PR. Can you tackle other sources of excessive connection establishment (e.g. cleanup)? Especially in the part where we block writes.

niupre commented:
@marcocitus, As mentioned in the Description, I will do that as part of taking over Jelte's PR.

marcocitus commented:
> @marcocitus, As mentioned in the Description, I will do that as part of taking over Jelte's PR.

sounds good, but I think that PR does not address things like cleanup

niupre commented:
@marcocitus, Yes, I understand. I will be adding that part in the PR.

onderkalaci commented on src/backend/distributed/operations/shard_cleaner.c:
```diff
@@ -75,7 +75,8 @@ static int DropOrphanedShardsForMove(bool waitForLocks);
 static bool TryDropShardOutsideTransaction(OperationId operationId,
```
I cannot see anything related to connection management here but few things to note for excessive connections:

- If you use `FORCE_NEW_CONNECTION` flag, as the name suggests, it'd establish a new connection.
- Instead, if you pass `connectionFlag = 0`, it'd try to use one of the existing connections to the given node, with the same user/database.
- At the end of the transaction, Citus closes all the connections, only keeps 1 (e.g., max_cached_conns_per_worker), see https://github.com/citusdata/citus/blob/3bf3d9c19b8631c6848f0c52865d1dad0e8693c4/src/backend/distributed/connection/connection_management.c#L1472-L1482



With this basic information out there, there are few strategies to avoid excessive connections:
- As @JelteF aimed here, keep the pointer to the connection and re-use: https://github.com/citusdata/citus/pull/6314/files
- Avoid FORCE_NEW_CONNECTION and let connection manager to pick the existing connection for you: https://github.com/citusdata/citus/commit/8be8d9ceb9bd127f9f5a34b3cb52c0e4090ecd42
- We definitely have a similar problem as above on TryDropShardOutsideTransaction. We should try re-use existiing

niupre replied:
Thanks Onder for the suggestion. The plan was to put the changes as part of the other PR.
I have removed all connection changes now and will continue on other PR.

onderkalaci reviewed (commented):
I have a strong objection to make "connection" related changes and "logging" changes in the same PR.

Connection changes are tricky and logging changes are crowded. So, we should tackle each separately

marcocitus commented on src/backend/distributed/operations/shard_cleaner.c:
```diff
@@ -490,15 +498,17 @@ FinalizeOperationNeedingCleanupOnFailure()
-							 removedShardCountOnComplete, list_length(
-								 currentOperationRecordList))));
+		ereport(LOG, (errmsg(
+						  "Processed %d orphaned shards out of %d as part of failed operation finalize ",
```
since this will also be exposed to users, "finalize" might be a bit confusing since it's an implementation (naming) detail

Maybe: "removed %d orphaned shards out of %d as part of a failed operations"

niupre replied:
We initially had verbiage like : "Removed 34 orphaned shards out of 67"
The suggestion was to disable the logging for cleanup of resources during the operation as it can be concerning : https://github.com/citusdata/citus/issues/6312.

This is why the Processed verbiage, do we want to go back to Remove?

niupre replied:
Fixed as per offline discussion.

marcocitus commented on src/backend/distributed/operations/shard_cleaner.c:
```diff
@@ -490,15 +498,17 @@ FinalizeOperationNeedingCleanupOnFailure()
-									 failedShardCountOnComplete, list_length(
-										 currentOperationRecordList))));
+			ereport(WARNING, (errmsg(
+								  "Unable to process %d shards out of %d as part of failed operation finalize",
```
maybe: "failed to clean up %d shards out of %d after a failed operation"

marcocitus commented on src/backend/distributed/operations/shard_cleaner.c:
```diff
@@ -572,15 +585,17 @@ FinalizeOperationNeedingCleanupOnSuccess()
-							 removedShardCountOnComplete, list_length(
-								 currentOperationRecordList))));
+		ereport(LOG, (errmsg(
+						  "Processed %d orphaned shards out of %d as part of operation finalize ",
```
similar feedback. I also liked the previous "Removed" wording a little more than "Processed".

niupre replied:
Hey @marcocitus , I thought the term "Removed" as part of inline cleaning in operation was "concerning" as you mentioned here : https://github.com/citusdata/citus/issues/6312. Why are we going back to it?

niupre replied:
Fixed as per offline discussion.

marcocitus commented on src/backend/distributed/operations/shard_split.c:
```diff
@@ -1638,10 +1682,16 @@ NonBlockingShardSplit(SplitOperation splitOperation,
 		 */
 		if (DeferShardDeleteOnSplit)
 		{
+			ereport(LOG, (errmsg("marking deferred cleanup of source shard(s) for %s",
```
thanks, these messages all seem really useful

marcocitus reviewed (approved):
Logging the individual steps is very nice and something we should do for more commands.

I did a few changes to cleanup logging, because I felt we could give nicer messages by pulling the logging out of TryDropShardOutsideTransaction and passing an operation name. It remains a bit tricky (ideally we could distinguish dummy shards, because they're a bit confusing), but this seems overall okay and seems like it will give sufficient information about what the system was doing when something goes wrong.

Operation that fails at the last moment:
```
2022-09-16 14:40:32.314 CEST [15888] LOG:  cleaned up orphaned shard public.test_102335 on localhost:1500 after a create_distributed_table_concurrently operation completed
2022-09-16 14:40:32.314 CEST [15888] STATEMENT:  select create_distributed_table_concurrently('test','x'), random()/0;
2022-09-16 14:40:32.327 CEST [15888] ERROR:  division by zero
2022-09-16 14:40:32.327 CEST [15888] STATEMENT:  select create_distributed_table_concurrently('test','x'), random()/0;
...
2022-09-16 14:40:46.098 CEST [15564] LOG:  cleaned up orphaned shard public.test_102336 on localhost:1500 which was left behind after a failed operation
2022-09-16 14:40:46.098 CEST [15564] CONTEXT:  Citus maintenance daemon for database 5 user 10
2022-09-16 14:40:46.101 CEST [15564] LOG:  maintenance daemon dropped 32 distributed shards previously marked to be removed
2022-09-16 14:40:46.101 CEST [15564] CONTEXT:  Citus maintenance daemon for database 5 user 10
```

Successful operation:
```
2022-09-16 14:39:16.468 CEST [15563] LOG:  cleaned up orphaned shard public.test_102300 on localhost:1500 after a create_distributed_table_concurrently operation completed
2022-09-16 14:39:16.468 CEST [15563] STATEMENT:  select create_distributed_table_concurrently('test','x');
2022-09-16 14:39:16.512 CEST [15563] LOG:  cleaned up orphaned shard public.test_102302 on localhost:1500 after a create_distributed_table_concurrently operation completed
2022-09-16 14:39:16.512 CEST [15563] STATEMENT:  select create_distributed_table_concurrently('test','x');
2022-09-16 14:39:25.263 CEST [15564] LOG:  deferred drop of orphaned shard public.test_102271 on localhost:1500 completed
2022-09-16 14:39:25.263 CEST [15564] CONTEXT:  Citus maintenance daemon for database 5 user 10
2022-09-16 14:39:25.267 CEST [15564] LOG:  maintenance daemon dropped 1 distributed shards previously marked to be removed
2022-09-16 14:39:25.267 CEST [15564] CONTEXT:  Citus maintenance daemon for database 5 user 10
```

niupre commented:
> Logging the individual steps is very nice and something we should do for more commands.
> 
> I did a few changes to cleanup logging, because I felt we could give nicer messages by pulling the logging out of TryDropShardOutsideTransaction and passing an operation name. It remains a bit tricky (ideally we could distinguish dummy shards, because they're a bit confusing), but this seems overall okay and seems like it will give sufficient information about what the system was doing when something goes wrong.
> 
> Operation that fails at the last moment:
> 
> ```
> 2022-09-16 14:40:32.314 CEST [15888] LOG:  cleaned up orphaned shard public.test_102335 on localhost:1500 after a create_distributed_table_concurrently operation completed
> 2022-09-16 14:40:32.314 CEST [15888] STATEMENT:  select create_distributed_table_concurrently('test','x'), random()/0;
> 2022-09-16 14:40:32.327 CEST [15888] ERROR:  division by zero
> 2022-09-16 14:40:32.327 CEST [15888] STATEMENT:  select create_distributed_table_concurrently('test','x'), random()/0;
> ...
> 2022-09-16 14:40:46.098 CEST [15564] LOG:  cleaned up orphaned shard public.test_102336 on localhost:1500 which was left behind after a failed operation
> 2022-09-16 14:40:46.098 CEST [15564] CONTEXT:  Citus maintenance daemon for database 5 user 10
> 2022-09-16 14:40:46.101 CEST [15564] LOG:  maintenance daemon dropped 32 distributed shards previously marked to be removed
> 2022-09-16 14:40:46.101 CEST [15564] CONTEXT:  Citus maintenance daemon for database 5 user 10
> ```
> 
> Successful operation:
> 
> ```
> 2022-09-16 14:39:16.468 CEST [15563] LOG:  cleaned up orphaned shard public.test_102300 on localhost:1500 after a create_distributed_table_concurrently operation completed
> 2022-09-16 14:39:16.468 CEST [15563] STATEMENT:  select create_distributed_table_concurrently('test','x');
> 2022-09-16 14:39:16.512 CEST [15563] LOG:  cleaned up orphaned shard public.test_102302 on localhost:1500 after a create_distributed_table_concurrently operation completed
> 2022-09-16 14:39:16.512 CEST [15563] STATEMENT:  select create_distributed_table_concurrently('test','x');
> 2022-09-16 14:39:25.263 CEST [15564] LOG:  deferred drop of orphaned shard public.test_102271 on localhost:1500 completed
> 2022-09-16 14:39:25.263 CEST [15564] CONTEXT:  Citus maintenance daemon for database 5 user 10
> 2022-09-16 14:39:25.267 CEST [15564] LOG:  maintenance daemon dropped 1 distributed shards previously marked to be removed
> 2022-09-16 14:39:25.267 CEST [15564] CONTEXT:  Citus maintenance daemon for database 5 user 10
> ```

Thanks @marcocitus!
